### PR TITLE
Greedy router

### PR DESCRIPTION
**Pull request by yourball**
Fixes issue #3056. Fixes hanging greedy router on some circuit instances by partially reverting changes from 6d81ddbb.

### CODE REVIEW

**Comment by yourball** [comment]
@googlebot I signed it!

**Review by balopat — CHANGES_REQUESTED** [review]
Thank you for taking the time digging this out!! The fix looks good to me, as it is covered by tests. However, it would be great to put a test around it so that we can prevent regression (like what we have now).

The test would run a minimal instance where we can prove that the algorithm would run on infinite loops without this fix, but fast (<1s) with the fix. The algorithm would run in a separate process that we should measure if it finishes within the expected time. Can you find a graph instance like that?

**Review by balopat — APPROVED** [review]
LGTM - One more nit 
Thanks for adding the test, I just tried it out and it works nicely! :) 
We should make the timeout 5 seconds to avoid flakes, on my Mac it runs just above 2 seconds.

File: cirq/contrib/routing/greedy_test.py [hunk 1]
```diff
@@ -20,10 +22,47 @@
...
+def test_router_hanging():
+    """Run a separate process and check if greedy router hits timeout (1s)."""
+    circuit, device_graph = create_circuit_and_device()
+    process = Process(target=create_hanging_routing_instance,
+                      args=[circuit, device_graph])
+    process.start()
+    process.join(timeout=1)
+    alive = process.is_alive()
+    if alive:
+        process.terminate()
```
Comment (balopat) [inline cirq/contrib/routing/greedy_test.py:67]: I would do something like this instead - this takes care of the coverage issue as well

```
    try:
        assert not process.is_alive(), "Greedy router timeout"
    finally:
        process.terminate()
```

Comment (yourball) [reply cirq/contrib/routing/greedy_test.py:68]: Cool, thanks! Was happy to contribute. It looks like all tests passed now.:)

**Pull request merged**